atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/pue/ResilientCommitTimestampPutUnlessExistsTable.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    @Override
+    public ListenableFuture<Map<Long, Long>> get(Iterable<Long> cells) {
+        Map<Long, Cell> startTsToCell = StreamSupport.stream(cells.spliterator(), false)
+                .collect(Collectors.toMap(x -> x, encodingStrategy::encodeStartTimestampAsCell));
+        ListenableFuture<Map<Cell, byte[]>> asyncReads = store.getMultiple(startTsToCell.values());
+        return Futures.transform(
+                asyncReads,
+                presentValues -> processReads(presentValues, startTsToCell),
+                MoreExecutors.directExecutor());
+    }
+
+    private byte[] pueStaging(Cell cell, long startTs, long commitTs) {
+        byte[] stagingValue =
+                encodingStrategy.encodeCommitTimestampAsValue(startTs, PutUnlessExistsValue.staging(commitTs));
+        store.putUnlessExists(cell, stagingValue);
+        return stagingValue;
+    }
+
+    private void putCommitted(Cell cell, byte[] stagingValue) {
+        store.put(cell, encodingStrategy.transformStagingToCommitted(stagingValue));
+    }
+
+    private Long processRead(Cell cell, Long startTs, Optional<byte[]> maybeActual) {
+        if (maybeActual.isEmpty()) {
+            return null;
+        }
+
+        byte[] actual = maybeActual.get();
+        PutUnlessExistsValue<Long> valueSeen = encodingStrategy.decodeValueAsCommitTimestamp(startTs, actual);
+
+        Long commitTs = valueSeen.value();
+        if (valueSeen.isCommitted()) {
+            return commitTs;
+        }
+        store.checkAndTouch(cell, actual);
+        putCommitted(cell, actual);
+        return commitTs;
+    }
+
+    private Map<Long, Long> processReads(Map<Cell, byte[]> reads, Map<Long, Cell> startTsToCell) {
+        Map<Cell, byte[]> checkAndTouch = new HashMap<>();
+        ImmutableMap.Builder<Long, Long> resultBuilder = ImmutableMap.builder();
+        for (Map.Entry<Long, Cell> startTsAndCell : startTsToCell.entrySet()) {
+            Cell cell = startTsAndCell.getValue();
+            Optional<byte[]> maybeActual = Optional.ofNullable(reads.get(cell));
+            if (maybeActual.isEmpty()) {
+                continue;
+            }
+
+            Long startTs = startTsAndCell.getKey();
+            byte[] actual = maybeActual.get();
+            PutUnlessExistsValue<Long> currentValue = encodingStrategy.decodeValueAsCommitTimestamp(startTs, actual);
+
+            Long commitTs = currentValue.value();
+            if (currentValue.isCommitted()) {
+                resultBuilder.put(startTs, commitTs);
+                continue;
+            }
+            checkAndTouch.put(cell, actual);
+            resultBuilder.put(startTs, commitTs);
+        }
+        store.checkAndTouch(checkAndTouch);
+        store.put(KeyedStream.stream(checkAndTouch)
+                .map(encodingStrategy::transformStagingToCommitted)
+                .collectToMap());
+        return resultBuilder.build();
+    }
```

Comment:
as elsewhere: how bad would it be to make processRead single call processReads? there's a fair bit of duplication currently